### PR TITLE
Refine job management component

### PR DESCRIPTION
**Pull request by hongyh13**
1. Refine the job management component.
2. Add more unit tests for job management.

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using Microsoft.Health.Fhir.Synapse.JobManagement.Extensions;
+using Microsoft.Health.Fhir.Synapse.JobManagement.Models;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Synapse.Core.Jobs.Models
+{
+    public class FhirToDataLakeAzureStorageJobInfo : AzureStorageJobInfo
+    {
+        private readonly DateTimeOffset _fakeDataEndTime = new DateTimeOffset(2022,1,1,0,0,0, TimeSpan.Zero);
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs:15]: Can we use DateTimeOffset.Min/Max to  avoid some magic numbers?

Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs:15]: Sure, updated.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs [hunk 2]
```diff
@@ -229,6 +241,8 @@ await foreach (var pageResult in jobLockEntityQueryResult.AsPages().WithCancella
                         jobLockEntity[JobLockEntityProperties.JobMessageId] = response.Value.MessageId;
                     }
 
+                    _logger.LogInformation($"[Enqueue] Send messages for jobs {string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())} successfully.");
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs:244]: nit, send queue message?

**Review by moria97 — APPROVED** [review]

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        public override string JobIdentifier()
+        {
+            string identifierString = Definition;
+            try
+            {
+                var inputData = JsonConvert.DeserializeObject<FhirToDataLakeOrchestratorJobInputData>(Definition);
+                if (inputData is { JobType: JobType.Orchestrator })
+                {
+                    inputData.DataEndTime = _fakeDataEndTime;
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs:25]: Sure, added.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                    identifierString = JsonConvert.SerializeObject(inputData);
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/Models/FhirToDataLakeAzureStorageJobInfo.cs:26]: Sure

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs (hunk 2 above)
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs:244]: Fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs [hunk 5]
```diff
@@ -72,6 +74,7 @@ public bool IsInitialized()
         // 3. if one of the steps fails, will continue to process it when re-enqueue.
         // 4. if there are multi jobs to be enqueued, all the job will operate in one transaction
         // TODO: The parameter forceOneActiveJobGroup and isCompleted are ignored for now
+        // Note: We don't allow the definitions to be enqueued partial overlap in different calls,  which means wo don't allow the first time to enqueue "job1" and "job2", and the second time to only enqueue "job1", or enqueue "job1" and "job3", or enqueue "job1", "job2" and "job3"
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs:77]: Yes.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageClientFactory.cs [hunk 6]
```diff
@@ -71,12 +74,12 @@ public class AzureStorageClientFactory : IAzureStorageClientFactory
         public TableClient CreateTableClient()
```
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageClientFactory.cs:74]: Sure, added.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hongyh13) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs]: Yes, the definitions are enqueued in one transaction, besides, the definitions in different calls should not partially overlap.

**Pull request merged**